README.md

#django-upthor

`django-upthor` provides a django application for simple ajax file uploads. We use
https://github.com/blueimp/jQuery-File-Upload for the upload functionality.

**Warning:** This isn't close to being a complete app, but it's getting there.


#Usage


##Step 1. Install

- `pip install django-upthor`

Now you have two options:

- If you want to encrypt FQ values, install pycrypto. `pip install pycrypto==2.6.1`
- Or you can, disable FQ encryption by adding `THOR_DISABLE_FQ_ENCRYPT = True` to your settings file.



##Step 2. (Django 1.6+)
Add 'upthor' to your installed apps in settings.py:

```
INSTALLED_APPS = (
    ...
    "upthor",
)
```

Then:

```
python manage.py migrate
```


##Step 3. Use it in your app's models.

```

import os
import uuid

from django.db import models
from upthor import fields as thor_fields


def random_upload_path(instance, filename):
    # Split the uuid into two parts so that we won't run into subdirectory count limits. First part has 3 hex chars,
    #  thus 4k possible values.
    uuid_hex = uuid.uuid4().hex
    return os.path.join(uuid_hex[:3], uuid_hex[3:], filename)


def post_example_file_link(real_instance, temporary_instance, raw_file):
    """
        A callback called after linking the temporary file with the model.
        
        **Warning**: Don't call instances save method from here, cause it will cause an recursion error.
    
        @:param real_instance An instance of the model the file is attached to
        @:param temporary_instance An instance of TemporaryFileWrapper that the form links to.
        @:param raw_file The raw file that is being uploaded.

        @:return bool If True, the uploaded temporary file is removed once the linking is complete.
    """
    return True


def get_file_image(file_path):
    """ An optional function that returns the display image html for files after uploading is complete"""
    
    return '<i class="fa fa-file"></i>'


class ExampleModelWithFile(models.Model):
    name = models.CharField(max_length=50)
    file = thor_fields.ThorFileField(upload_to=random_upload_path,
                                   allowed_types=['*'], widget=thor_fields.ThorSingleUploadWidget,
                                   post_link=post_product_file_link,
                                   get_upload_image=get_file_image)
```


##Step 4. Make sure to include form media.

Make sure you include the media files for the form in your templates:

E.g. Add the following codes where form is the context 
object of your modelform that uses the uploader fields.

```
    {{ form.media.css }}
    
    {{ form.media.js }}
```


##Step 5. Add the upload url to your project urls.

```
    url(r'', include('upthor.urls')),
```


##Step 6. Optional stuff

#### Temporary file cleanup

If you want to clean up temporary files automatically, you'll need to install [django-cron](https://github.com/Tivix/django-cron) and add `upthor.cron.CleanTemporaryFiles` to your cron classes in settings.

Alternatively to clean up manually you can use the management command `clean_temporary_files`.

#### Custom upload widget template

You can override `ThorSingleUploadWidget.render_template` to return your own widget template instead of the [hardcoded one defined in widgets.py](upthor/widgets.py). Although the structure (including most classes) has to remain the same, there are a few data attributes on `.file-upload` that you can use to customize behavior:

| Data Attribute Name | Type    | Description                              |
| ------------------- | ------- | ---------------------------------------- |
| upload-url          | string  | **Required:** URL to POST temporary files to, defaults to reverse of `thor-file-upload`. |
| max-size            | number  | **Required:** Maximum allowed file size in bytes, defaults to `THOR_MAX_FILE_SIZE`. |
| size-error          | string  | **Required:** Text to display if the file doesn't meet the size requirements, defaults to ` "Uploaded file too large"`. |
| use-background      | boolean | Whether or not to use `background-image` instead of `img` elements, defaults to false. |


#Backends

Currently it only supports local file backend, but we plan to add other backends when we reach a stable state.


#Settings

The following settings are customizable using your django project settings file.

**THOR_UPLOAD_TO**

Path where the upload files will be stored. Defaults to "temp-files".

**THOR_EXPIRE_TIME**

How long to keep temporary files in the database and on disk. Defaults to "60*60*24", e.g. 24 hours.

**THOR_LINKED_EXPIRE_TIME**

How long to keep linked temporary files in the database and on disk. Defaults to "60*60*6", e.g. 6 hours.

**THOR_MAX_FILE_SIZE**

The max file size of uploaded files. Defaults to "2*1024*1024", e.g. 2 MB. 

**THOR_DISABLE_FQ_ENCRYPT**

Disable the FQ Encryption, if this is False you need to install pycrypto since that is used for encryption. Defaults to "False".

**THOR_ENABLE_ADMIN**

Should TemporaryFileWrapper model be shown in the admin interface. Defaults to "True".


CHANGELOG.md
## Changelog

#### 0.9.1

Initial PyPI release

Changes:

- Updated to be compatible with Django >= 1.6, **dropped support for older django versions**
- Updated jQuery UI
- Added optional `data-use-background` to use `background-image` instead of `img`
- Added optional `data-find-controls-element` to specify a custom controls container

#### 0.1.43

Initial alpha GitHub release

upthor/models.py
import base64
import hashlib
import os
import uuid

from django.conf import settings
from django.db import models
from django.db.models.signals import post_delete
from django.dispatch import receiver
from django.template.defaultfilters import filesizeformat
from django.utils.encoding import force_text, force_bytes
from django.utils.translation import ugettext_lazy as _


def get_upload_path():
    return getattr(settings, 'THOR_UPLOAD_TO', 'temp-files')


def get_expiry_time():
    return getattr(settings, 'THOR_EXPIRE_TIME', 60*60*24)


def get_linked_expiry_time():
    return getattr(settings, 'THOR_LINKED_EXPIRE_TIME', 60*60*6)


def get_max_file_size():
    return getattr(settings, 'THOR_MAX_FILE_SIZE', 2*1024*1024)


def fq_encrypt_disabled():
    return getattr(settings, 'THOR_DISABLE_FQ_ENCRYPT', False)


def show_in_admin():
    return getattr(settings, 'THOR_ENABLE_ADMIN', True)


def get_size_error():
    return force_text(_("Uploaded file too large ( > %s )") % filesizeformat(get_max_file_size()))


def thor_upload_file_name(instance, filename):
    if len(filename) > 40:
        filename = filename[-40:]

    uuid_hex = uuid.uuid4().hex
    # This removes "False" from the filename, which helps us to distinct between empty values in widgets. :)
    filename = filename.replace('False', uuid_hex[3:8])
    return os.path.join(get_upload_path(), uuid_hex[:3], uuid_hex[3:], filename)


def human_readable_types(types):
    ret = ['.%s' % (x.split('/')[-1] if "/" in x else x) for x in types]
    return ', '.join(ret)


class TemporaryFileWrapper(models.Model):
    """ Holds an arbitrary file and notes when it was last accessed
    """
    file = models.FileField(upload_to=thor_upload_file_name)

    modified = models.DateTimeField(auto_now=True)
    md5sum = models.CharField(max_length=36, unique=True)

    content_type = models.CharField('content_type', max_length=128, default='application/unknown')
    linked = models.BooleanField(default=False)

    def __str__(self):
        return '%s%s' % (
            '[LINKED] ' if self.linked else '',
            self.file.name,
        )

    def get_hash(self):
        md5 = hashlib.md5()
        for chunk in self.file.chunks():
            md5.update(chunk)

        return md5.hexdigest()

    def save(self, force_insert=False, force_update=False, using=None, update_fields=None):
        self.md5sum = self.get_hash()

        if not self.pk or TemporaryFileWrapper.objects.exclude(id=self.pk).filter(md5sum=self.md5sum).exists():
            try:
                # If replacing pk, we mark the file as unlinked.
                self.linked = False
                self.pk = TemporaryFileWrapper.objects.exclude(id=self.pk).get(md5sum=self.md5sum).pk
            except TemporaryFileWrapper.DoesNotExist:
                pass

        return super(TemporaryFileWrapper, self).save(force_insert, force_update, using, update_fields)

    @staticmethod
    def get_image_from_id(img_id, field_query):
        try:
            img_id = int(img_id)
        except (ValueError, TypeError):
            return None
        else:
            model = field_query[0]

            try:
                instance = model.objects.get(id=img_id)
            except model.DoesNotExist:
                return None
            else:
                return getattr(instance, field_query[1])


@receiver(post_delete, sender=TemporaryFileWrapper)
def cleanup_temporary_files(sender, instance, **kwargs):
    instance.file.close()
    storage, path = instance.file.storage, instance.file.path
    try:
        storage.delete(path)
    except IOError:
        pass


class FqCrypto(object):
    BLOCK_SIZE = 32
    PADDING = '{'

    @classmethod
    def _pad(cls, s):
        return s + (cls.BLOCK_SIZE - len(s) % cls.BLOCK_SIZE) * cls.PADDING

    @classmethod
    def _encode_aes(cls, c, s):
        return base64.b64encode(c.encrypt(cls._pad(s)))

    @classmethod
    def _decode_aes(cls, c, e):
        e = force_bytes(e, 'utf-8')
        pad = force_bytes(cls.PADDING, 'utf-8')

        return c.decrypt(base64.b64decode(e)).rstrip(pad)

    @classmethod
    def _cipher(cls):
        secret = settings.SECRET_KEY

        if len(secret) < cls.BLOCK_SIZE:
            secret = cls._pad(secret)
        else:
            secret = secret[:cls.BLOCK_SIZE]

        from Crypto.Cipher import AES
        return AES.new(secret)

    @classmethod
    def decode(cls, value):
        if fq_encrypt_disabled():
            return force_text(value)

        return force_text(cls._decode_aes(cls._cipher(), value))

    @classmethod
    def encode(cls, value):
        if fq_encrypt_disabled():
            return value

        return force_text(cls._encode_aes(cls._cipher(), value))
